### PR TITLE
Allow a counter to be manually reset

### PR DESCRIPTION
**Pull request by nullobject**
- [x] Depends on #1515.

This PR adds a `reset` parameter to the Counter API, that allows a counter to be reset to its initial value.

It is sometimes useful to be able to reset a counter arbitrarily. By default the `reset` parameter is `false`, which means the proposed change is compatible with the existing API.

For example

```scala
val counterReset = WireInit(false.B)
val (value, wrap) = Counter(0 to 3, reset = counterReset)

// Trigger a reset
counterReset := true.B
```

**Type of change**: other enhancement

**Impact**: API addition (no impact on existing code)

**Development Phase**: implementation

**Release Notes**
Allow counters to be reset

### CODE REVIEW

**Comment by seldridge** [comment]
With #1515 merged, would you be able to rebase this onto `master` @nullobject?

And thanks for this. It's something that I've written one-off workarounds on multiple occasions for.

**Comment by nullobject** [comment]
@seldridge Sure thing, I will get it out of draft status and mergable :+1:

**Comment by nullobject** [comment]
> With #1515 merged, would you be able to rebase this onto master @nullobject?

@seldridge It has been rebased.

File: src/test/scala/chiselTests/Counter.scala [hunk 1]
```diff
@@ -57,6 +70,10 @@ class CounterSpec extends ChiselPropSpec {
     forAll(safeUInts) { (seed: Int) => whenever(seed >= 0) { assertTesterPasses{ new EnableTester(seed) } } }
   }
 
+  property("Counter can be reset") {
+    forAll(smallPosInts) { (seed: Int) => assertTesterPasses{ new ResetTester(seed) } }
+  }
+
```
Comment (seldridge) [inline src/test/scala/chiselTests/Counter.scala:76]: This test is probably overkill and it seems to repeat (at least locally, for me) seeds. Out of scope for this PR, though. I like having this PR consistent with the other tests here.

**Pull request merged**